draft-iyengar-quic-delayed-ack.md

```diff
@@ -288,15 +287,23 @@ send an acknowledgement immediately on receiving a reordered ack-eliciting
 packet. This extension modifies this behavior.
 
 If the endpoint has not yet received an ACK_FREQUENCY frame, or if the most
-recent frame received from the peer has an `Ignore Order` value of `false`
-(0x00), the endpoint MUST immediately acknowledge any subsequent packets that
-are received out of order.
-
-If the most recent ACK_FREQUENCY frame received from the peer has an `Ignore
-Order` value of `true` (0x01), the endpoint does not make this exception. That
-is, the endpoint MUST NOT send an immediate acknowledgement in response to
-packets received out of order, and instead continues to use the peer's `Packet
-Tolerance` and `max_ack_delay` thresholds for sending acknowledgements.
+recently processed ACK_FREQUENCY frame has a `Packet Loss Threshold` value of 1,
+the endpoint MUST immediately acknowledge any subsequent packets that are
+received out of order.
+
+If the most recently processed ACK_FREQUENCY frame has a 'Packet Loss Threshold'
+larger than 1, implementations MUST still immediately acknowledge any received
+packets with packet numbers smaller than the largest 'Largest Acked' value
+previously sent in an ACK frame.  If the packet number is larger than the largest
+'Largest Acked' and there are missing packets that are at least
+'Packet Loss Threshold' smaller than the largest received packet, an
+immediate acknowledgement MUST be sent.
```

Comment:
I don't know how to implement this.  Or maybe it's just a bug.

Let's say that I acknowledged packet 10.  The packet loss threshold is 4.  As stated, this says: acknowledge immediately if you receive packet 15.  That seems OK.

But if I acknowledge 10, but 9 has not been received, this text implies that receiving 13 would cause an immediate ACK.  That seems like it might always result in an immediate ACK over time:

```
if pn > largest_received:
  largest_received = pn
  for n in 0..(largest_received - loss_threshold):
    if !received(n):
      send_immediate_ack()
      break
```

If instead this is something like:

```
if pn > largest_received:
  largest_received = pn
  if loss_threshold > 0 && largest_received > largest_acked + loss_threshold:
    for n in (largest_acked + 1)..(largest_received - loss_threshold):
      if !received(n):
        send_immediate_ack()
        break
```

That might be a reason to include the largest acknowledged in the calculation.

Assuming the latter is your intent, as this requires iteration over `loss_threshold` packets, receivers should be allowed to limit the size of this value arbitrarily to avoid O(N) work for little attacker cost.

Reply:
FWIW, I tend to think that the latter is the correct thing to do. But at the same time, I would argue against allowing the receiver to set an arbitrary maximum for the loss_threshold field, as that prevents the sender from using a loss_threshold as a way of asking the receiver to ignore order.

In practice, I would assume that a receiver would be using a list of ranges (i.e. ack blocks) for representing what needs to be sent in an ACK frame. Typically, the number of blocks will exceed an internal threshold and trigger an ack before the attack being described here becomes an issue. I would be fine with pointing that out explicitly, and that would work as a mitigation.

Reply:
As a followup to this.  I tried implementing this and realized that if you hold ranges of packets, you can do something sensible without having to limit the value of `loss_threshold`, but it isn't very easy to think about.

If you haven't sent an ACK yet, then you can check that the `range.min > 0` and `range.max > loss_threshold`.  This says that the range of unacknowledged packets you have received doesn't start at 0 and there might be more than `loss_threshold` packets outstanding.  This bit gets a little sketchy when people start packet numbers at numbers greater than 0, but I think that the immediate ACK there is tolerable in that case, because it happens just once when it might not be desirable (that is, for the first application data packet).

If you have sent an ACK including `largest`, then you check that `range.min > largest + 1` and `range.max > largest + loss_threshold`.

This only works if you merge ranges properly when receiving packets.  For instance, when receiving packets 3, 1, and 2 in that order, you produce a single range that covers 1 to 3 and not separate ranges.  Merging ranges is fairly straightforward, and you more or less have to it to construct valid ACK frames, but it's worth noting.

I ultimately track the *next* packet that needs acknowledgment rather than the largest (this saves errors at 0), and so I get:

```
on_received(pn):

ranges.add(pn)
if pn == ranges[0].max {
  # this just extended the most recent range
  last_packet_time = now()
  if loss_threshold > 0:
    reordered = next_unacknowledged < ranges[0].min && pn >= next_unacknowledged + loss_threshold
  else:
    reordered = false
else:
  reordered = pn < next_unacknowledged
endif
```

Reply:
The way I'd solve this is to look up the oldest unacked packet.  This packet would have the smallest packet number.  This is the packet number to subtract from `largest_received`; the difference is to be compared with `loss_threshold`.